### PR TITLE
[statedb] use read lock to access currentChainHeight in reader funcs

### PR DESCRIPTION
**Pull request by dustinxie**
currentChainHeight is only written in `ws.Commit(ctx)`, which is protected by a write lock: https://github.com/iotexproject/iotex-core/blob/master/state/factory/statedb.go#L424-L440 
for all other places, read lock should suffice

test coverage:
1. make test
2. verified on full-node it can successfully commit/sync blocks

### CODE REVIEW

File: state/factory/statedb.go [hunk 1]
```diff
@@ -553,22 +556,20 @@ func (sdb *stateDB) createGenesisStates(ctx context.Context) error {
...
+	sdb.mutex.RLock()
+	h := sdb.currentChainHeight
+	sdb.mutex.RUnlock()
+	tx, err := sdb.newWorkingSet(ctx, h+1)
 	return tx, false, err
 }
 
 func (sdb *stateDB) putIntoWorkingSets(key hash.Hash256, ws *workingSet) {
-	sdb.mutex.Lock()
-	defer sdb.mutex.Unlock()
```
Comment (dustinxie) [inline state/factory/statedb.go:572]: sdb.workingsets.Add() is thread-safe, no need to use lock here

File: state/factory/statedb.go [hunk 2]
```diff
@@ -318,10 +318,11 @@ func (sdb *stateDB) NewBlockBuilder(
 	ap actpool.ActPool,
 	sign func(action.Envelope) (action.SealedEnvelope, error),
 ) (*block.Builder, error) {
-	sdb.mutex.Lock()
 	ctx = protocol.WithRegistry(ctx, sdb.registry)
-	ws, err := sdb.newWorkingSet(ctx, sdb.currentChainHeight+1)
-	sdb.mutex.Unlock()
+	sdb.mutex.RLock()
+	h := sdb.currentChainHeight
```
Comment (Liuhaai) [inline state/factory/statedb.go:323]: Replacing single char var `h` with longer var like `chainHeight` will make code editor easy to locate the var

File: state/factory/statedb.go [hunk 3]
```diff
@@ -553,22 +556,20 @@ func (sdb *stateDB) createGenesisStates(ctx context.Context) error {
...
+	sdb.mutex.RUnlock()
+	tx, err := sdb.newWorkingSet(ctx, h+1)
 	return tx, false, err
 }
 
 func (sdb *stateDB) putIntoWorkingSets(key hash.Hash256, ws *workingSet) {
-	sdb.mutex.Lock()
-	defer sdb.mutex.Unlock()
 	sdb.workingsets.Add(key, ws)
```
Comment (Liuhaai) [inline state/factory/statedb.go:574]: This code is very straightforward. I think this func can be removed.

**Review by Liuhaai — APPROVED** [review]

File: state/factory/statedb.go [hunk 4]
```diff
@@ -446,27 +449,27 @@ func (sdb *stateDB) DeleteTipBlock(_ *block.Block) error {
...
 	cfg, err := processOptions(opts...)
 	if err != nil {
 		return 0, err
 	}
-
-	return sdb.currentChainHeight, sdb.state(cfg.Namespace, cfg.Key, s)
+	sdb.mutex.RLock()
+	h := sdb.currentChainHeight
+	sdb.mutex.RUnlock()
+	return h, sdb.state(cfg.Namespace, cfg.Key, s)
```
Comment (CoderZhi) [inline state/factory/statedb.go:459]: state will read data from dao and then deserialize it. If we don't lock the whole struct, will `h` be inconsistent with the result read from `sdb.state`?

File: state/factory/statedb.go [hunk 5]
```diff
@@ -446,27 +449,27 @@ func (sdb *stateDB) DeleteTipBlock(_ *block.Block) error {
...
 func (sdb *stateDB) States(opts ...protocol.StateOption) (uint64, state.Iterator, error) {
-	sdb.mutex.RLock()
-	defer sdb.mutex.RUnlock()
 	cfg, err := processOptions(opts...)
 	if err != nil {
 		return 0, nil, err
 	}
+	sdb.mutex.RLock()
+	h := sdb.currentChainHeight
+	sdb.mutex.RUnlock()
```
Comment (CoderZhi) [inline state/factory/statedb.go:470]: same question for `sdb.dao.Filter`

File: state/factory/statedb.go (hunk 2 above)
Comment (dustinxie) [reply state/factory/statedb.go]: fixed

File: state/factory/statedb.go (hunk 3 above)
Comment (dustinxie) [reply state/factory/statedb.go]: fixed

File: state/factory/statedb.go (hunk 4 above)
Comment (dustinxie) [reply state/factory/statedb.go]: fixed

File: state/factory/statedb.go (hunk 5 above)
Comment (dustinxie) [reply state/factory/statedb.go]: fixed

**Review by CoderZhi — APPROVED** [review]
The change won't have significant impact on the performance.

**Pull request merged**